cli/listener/src/main.rs

```diff
@@ -0,0 +1,190 @@
+    /// Unix timestamp when the data point was logged.
+    timestamp: u64,
+    /// Slot when the data point was logged.
+    slot: Slot,
+    /// Epoch when the data point was logged.
+    epoch: Epoch,
+    /// Pool identifier, e.g. for Solido would be "solido".
+    pool: String,
+    token_a: u64,
+    token_b: u64,
+}
+
+pub fn create_db(conn: &Connection) -> rusqlite::Result<()> {
+    conn.execute(
+        "CREATE TABLE IF NOT EXISTS exchange_rate (
+                timestamp   INTEGER PRIMARY KEY,
+                slot        INTEGER NOT NULL,
+                epoch       INTEGER NOT NULL,
+                pool        TEXT NOT NULL,
+                token_a     INTEGER NOT NULL,
+                token_b     INTEGER NOT NULL
+            )",
+        [],
+    )?;
+    Ok(())
+}
+
+pub fn get_average_apy(conn: &Connection, opts: &Opts) -> rusqlite::Result<Option<f32>> {
+    let row_map = |row: &Row| {
+        Ok(ExchangeRate {
+            timestamp: row.get(0)?,
+            slot: row.get(1)?,
+            epoch: row.get(2)?,
+            pool: row.get(3)?,
+            token_a: row.get(4)?,
+            token_b: row.get(5)?,
+        })
+    };
+
+    // Get first logged entry by timestamp.
+    let mut stmt_first = conn.prepare(
+            "SELECT timestamp, slot, epoch, pool, token_a, token_b FROM exchange_rate WHERE pool = :pool ORDER BY timestamp ASC LIMIT 1",
+        )?;
+    let mut row_iter = stmt_first.query_map([opts.pool.clone()], row_map)?;
+    let first = row_iter.next();
+
+    // Get last logged entry by timestamp.
+    let mut stmt_last = conn.prepare(
+            "SELECT timestamp, slot, epoch, pool, token_a, token_b FROM exchange_rate WHERE pool = :pool ORDER BY timestamp DESC LIMIT 1",
+        )?;
+    let mut row_iter = stmt_last.query_map([opts.pool.clone()], row_map)?;
+    let last = row_iter.next();
+
+    match (first, last) {
+        (Some(first), Some(last)) => {
+            let exchange_rate_first = first?;
+            let exchange_rate_last = last?;
+            // Not enough data, need at least two data points.
+            if exchange_rate_first.timestamp == exchange_rate_last.timestamp {
+                Ok(None)
+            } else {
+                let seconds_in_year = 31536000f32;
+                // Will not underflow because we order them in the query.
+                let duration = exchange_rate_last.timestamp - exchange_rate_first.timestamp;
+                let fraction_of_year = duration as f32 / seconds_in_year;
+                let p0 = exchange_rate_first.token_a as f32 / exchange_rate_first.token_b as f32;
+                let p1 = exchange_rate_last.token_a as f32 / exchange_rate_last.token_b as f32;
+                let apy = (p1 / p0).powf(1. / (fraction_of_year));
+                Ok(Some((apy - 1f32) * 100f32))
+            }
+        }
+        _ => Ok(None),
+    }
+}
+
+pub fn log_price(conn: &Connection, exchange_rate: ExchangeRate) -> rusqlite::Result<()> {
```

Comment:
Nit: "log" sounds to me like it logs something to a log file or stdout/stderr, I would call this `insert_price`.